### PR TITLE
Faster attestation production

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

- Makes attestation production orders of magnitude faster in the best-case scenario.
- Greatly improves speed for getting shuffling at future epochs by introducing `StateSkipConfig` which allows for skipping forward without calculating state roots.
- Adds some new metrics for tracking key API response times.

## TODO

- [x] Address `SlotOutOfBounds` error.

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -578,7 +605,12 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         let mut state = if epoch == as_epoch(head_state.slot) {
             self.head()?.beacon_state
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Good spotting! Deleted!

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 2]
```diff
@@ -815,16 +834,23 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                     .ok_or_else(|| Error::MissingBeaconState(attestation_block_root))?
             } else {
                 let mut state = self
-                    .store
-                    .get_state(&attestation_block_root, Some(attestation_head_block.slot()))?
-                    .ok_or_else(|| Error::MissingBeaconState(attestation_block_root))?;
+                    .get_state_caching_only_with_committee_caches(
+                        &attestation_head_block.state_root(),
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: This will get replaced in #841, perhaps we can leave it for now? :)

**Comment by paulhauner** [comment]
All comments addressed :)

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 3]
```diff
@@ -563,154 +590,104 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
-        if let Some(attestation_duty) =
-            state.get_attestation_duties(validator_index, RelativeEpoch::Current)?
-        {
-            Ok(Some((attestation_duty.slot, attestation_duty.index)))
-        } else {
-            Ok(None)
-        }
-    }
-
-    /// Produce an `Attestation` that is valid for the given `slot` and `index`.
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: Comment deleted accidentally?

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Yup!

[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Sounds good

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 3 above)
Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: Thank you!

Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: (Fixed)

**Pull request merged**